File: WebAPI/Services/VoucherService.cs
using WebAPI.Domain;
using WebAPI.Dto;
using WebAPI.Repositories;

namespace WebAPI.Services
{
    public class VoucherService : IVoucherService
    {
        private readonly IVoucherRepository _voucherRepository;

        public VoucherService(IVoucherRepository voucherRepository)
        {
            _voucherRepository = voucherRepository;
        }

        public List<Voucher> GetVouchers()
        {
            return _voucherRepository.GetAll();
        }

        public int CreateVoucher(VoucherDto voucher )
        {
            if (voucher == null )
            {
                throw new Exception( $"{nameof( Voucher )} is not found!" );
            }

            Voucher voucherEntity = voucher.ConvertToVoucher();

            return _voucherRepository.Create(voucherEntity);
        }

        public void DeleteVoucher( int voucherId )
        {
            Voucher voucher = _voucherRepository.GetById( voucherId );
            if (voucher == null )
            {
                throw new Exception( $"{nameof(Voucher)} not found, #Id - {voucherId}" );
            }

            _voucherRepository.Delete(voucher);
        }

        public Voucher GetVoucherById( int voucherId )
        {
            Voucher voucher = _voucherRepository.GetById(voucherId);
            if (voucher == null )
            {
                throw new Exception( $"{nameof(Voucher)} not found, #Id - {voucherId}" );
            }
            return voucher;
        }


        public int UpdateVoucher(VoucherDto voucher)
        {
            if (voucher == null)
            {
                throw new Exception($"{nameof(Voucher)} is not found!");
            }

            return _voucherRepository.Update(voucher.ConvertToVoucher());
        }

    }
}

File: ToDoList.Site/ClientApp/src/app/travell-list/shared/voucher.service.ts
import { Injectable } from "@angular/core";
import { IVoucher } from './voucher.interface';
import { Observable } from "rxjs";
import { HttpClient } from "@angular/common/http";

@Injectable()
export class VoucherService {
    private readonly apiUrl: string = 'http://localhost:4200/rest/Voucher';

    constructor(private httpClient: HttpClient) {
    }

    public addVoucher(voucher: IVoucher): Observable<null> {
        return this.httpClient.post<null>(`${this.apiUrl}/create`, voucher);
    }

    public getVouchers(): Observable<IVoucher[]> {
        return this.httpClient.get<IVoucher[]>(this.apiUrl);
    }

    public getVoucherById(id: number): Observable<IVoucher[]> {
        return this.httpClient.get<IVoucher[]>(`${this.apiUrl}/${id}`);
    }

    public updateVoucher(voucher: IVoucher): Observable<null> {
        return this.httpClient.put<null>(`${this.apiUrl}/update`, voucher);
    }

    public deleteVoucher(id: number): Observable<object> {
        return this.httpClient.delete(`${this.apiUrl}/${id}/delete`);
    }
}